ioi-09-mecho.mdx

```diff
@@ -0,0 +1,150 @@
+of the four nodes surrounding his cave. If he did, then Mecho successfully reached 
+the cave while eating for x time. Otherwise, Mecho will have to eat for less than x 
+time.
+
+Output the maximum waiting time found out by the binary search. If Mecho cannot reach
+the cave even when he waited for 0 units of time, then there is no path from his 
+starting position to the cave. In this case, output -1.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^2\log^2N^2)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int MAX_N = 800;
+vector <string> field(MAX_N);
+bool bees_visited[MAX_N][MAX_N], mecho_visited[MAX_N][MAX_N];
+//mecho_time records the time taken for mecho to reach node[x][y].
+//bees_time does the same for the bees
+int mecho_time[MAX_N][MAX_N], bees_time[MAX_N][MAX_N];
+queue <pair <int, int>> q;
+int dx[] = {-1, 1, 0, 0};
+int dy[] = {0, 0, -1, 1};
+int n, s;
+bool valid_sq(int x, int y){
+    if (x >= 0 && x < n && y >= 0 && y < n && (field[x][y] == 'G' || field[x][y] == 'M'))
+        return true;
+    return false;
+}
+int main()
+{
+    cin >> n >> s;
+    for (int i = 0; i < n; i++)
+        cin >> field[i];
+    vector <pair <int, int>> hives;
+    int mechox, mechoy, home_x, home_y;
+    //find x and y coordinates for for Mecho, the bees and the cave
+    for (int i = 0; i < n; i++){
+        for (int j = 0; j < n; j++){
+            if (field[i][j] == 'M'){
+                mechox = i;
+                mechoy = j;
+            }else if (field[i][j] == 'H'){
+                hives.push_back({i, j});
+            }else if (field[i][j] == 'D'){
+                home_x = i;
+                home_y = j;
+            }
+        }
+    }
+    //binary search on waiting time
+    int l = 0, r = n*n;
+    while (l <= r){
+        int eating_time = (l + r) / 2;
+        memset(bees_visited, false, sizeof(bees_visited));
+        memset(mecho_visited, false, sizeof(mecho_visited));
+        memset(bees_time, 0, sizeof(bees_time));
+        memset(mecho_time, 0, sizeof(mecho_time));
+        //move bees
+        for (auto i : hives){
+            q.push({i.first, i.second});
+            bees_visited[i.first][i.second] = true;
+        }
+        while (!q.empty()){
+            int x = q.front().first, y = q.front().second;
+            q.pop();
+            for (int i = 0; i < 4; i++){
+                int nx = x + dx[i], ny = y + dy[i];
+                if (valid_sq(nx, ny) && !bees_visited[nx][ny]){
+                    bees_time[nx][ny] = bees_time[x][y] + 1;    
+                    q.push({nx, ny});
+                    bees_visited[nx][ny] = true;
+                }
+            }
+        }
+        //move Mecho
+        q.push({mechox, mechoy});
+        mecho_visited[mechox][mechoy] = true;
+        if (bees_time[mechox][mechoy] <= eating_time)
+            q.pop();
+        while (!q.empty()){
```

Comment:
tabs

Reply:
this hasn't been addressed. see https://usaco.guide/general/adding-solution?#general (2nd bullet point)